README.md
# Spatial-KWD

GitHub actions: 

[![build-update Actions Status](https://github.com/eurostat/Spatial-KWD/workflows/build-update/badge.svg)](https://github.com/eurostat/Spatial-KWD/actions) 

Downloads of the Python wrapper from PyPI:

<a class="reference external" href="https://badge.fury.io/py/Spatial-KWD"><img alt="PyPI version" src="https://badge.fury.io/py/Spatial-KWD.svg"></a> [![Downloads](https://pepy.tech/badge/spatial-kwd)](https://pepy.tech/project/spatial-kwd) [![Downloads](https://pepy.tech/badge/spatial-kwd/month)](https://pepy.tech/project/spatial-kwd) [![Downloads](https://pepy.tech/badge/spatial-kwd/week)](https://pepy.tech/project/spatial-kwd)

Downloads of the R package from CRAN: 

![](http://cranlogs.r-pkg.org/badges/grand-total/SpatialKWD) ![](http://cranlogs.r-pkg.org/badges/last-week/SpatialKWD)


## <a name="Description"></a>Computing Kantorovich-Wasserstein distances for large spatial maps

This software library contains efficient implementations of Discrete Optimal Transport algorithms for the computation of Kantorovich-Wassestein distances [1] customized for large spatial maps.

The core library is written in standard ANSI-C++11, but it has two wrappers:

1. A [Python wrapper](https://pypi.org/project/Spatial-KWD/) available from PyPI
2. An [R wrapper](https://CRAN.R-project.org/package=SpatialKWD) available from CRAN

If you need a wrapper in another language, please let us know by posting a request on GitHub.

Currently, the Spatial-KWD library is tested on

* Google Colabs notebooks
* Windows 10 (R v4.0.5, Python >= 3.6)
* Mac OS X Bug Sur 11.0.1 (R v4.0.5, Python 3.8.3)
* Linux 20.04.1 LTS (R v4.0.0, Python 3.8.5)
* Solaris 10 (R v4.0.5)

## <a name="Requirements"></a>Basic Usage: Colab Notebooks

The simplest way to test this library is to run one of the following notebooks on Colab, which include an example of installation of the library.

| Data | Notebook | Link |
|:-|:-|:-|
|**[2021/05/10]**|*Tutorial 1: Using Spatial-KWD with Python*|[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/eurostat/Spatial-KWD/blob/main/notebooks/Spatial_KWD_Tutorial_1.ipynb)|
|**[2020/05/10]**|*Tutorial 2: Using Spatial-KWD with R*|[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/eurostat/Spatial-KWD/blob/main/notebooks/Spatial_KWD_with_R_Tutorial_2.ipynb)|


## <a name="Python-wrapper"></a>Python Wrapper

The simplest way to install the python wrapper is to run the following command:

```
pip install Spatial-KWD
```

For basic usage of the library, please, look at the Colab notebok.

To compile the [Python](https://www.python.org/) wrapper directly from the source code, you need the following library:

* [Cython](https://cython.org/) (>= 0.23)

To install Cython you can look at the official documentation [Installing Cython](https://cython.readthedocs.io/en/latest/src/quickstart/install.html).

Then, you have to download this repository, and run from command line the following command:

```
make buildpython
```

This will compile the C++ code and build the python wrapper. If you prefer to compile the wrapper in a local directory, you can run the command:

```
python setup.py build_ext --inplace
```

In this case, however, you can only use the library in the local directory where you compile it.


## <a name="R-wrapper"></a>R Wrapper

The simplest way to install the R wrapper is to run the following command from the R shell:

```
install.packages("SpatialKWD")
```

For basic usage of the library, please, look at the Colab notebok.

To compile the wrapper directly from the source code, you need a recent version of [R](https://www.r-project.org/) and the [Rcpp](https://cran.r-project.org/web/packages/Rcpp/index.html) package (we test it with Rcpp v.1.0.5, but it should work also with older versions).

Windows users can download a pre-compiled binary package at the following link:

* [SpatialKWD_0.4.1.zip](https://github.com/eurostat/Spatial-KWD/releases/download/v0.4.1-alpha/SpatialKWD_0.4.1.zip)


If you need an interface for a different R data structure, please, drop us an email.


## <a name="Cpp-CLI"></a>Cpp Command Line Tool

By compiling the Cpp source code, it is possible to use the **SpatialKWD** directly from the command line.
The compilation is ruled by a simple Makefile, and is executed by the command:

```
make laptop
```

This build an executable named `solver` in the `bin` folder, which can be tested with the following command, which use a synthetic data from the whole Belgium region:

```
./bin/solver data/L_1000_100.csv
```

The output shold be as follows:

```
data/L_1000_100.csv
start solver
WARNING: the Xs input coordinates are not consecutives integers.
WARNING: the Ys input coordinates are not consecutives integers.
INFO: Recoding the input coordinates to consecutive integers.
INFO: change <verbosity> to info
INFO: change <opt_tolerance> to 0.000001
Approx => 32140: fobj: 0.689998, time: 2.1860, status: Optimal, iter: 167063, arcs: 58987, nodes: 39302
```

### <a name="About"></a>About

<table align="center">
    <tr> <td align="left"><i>Contributors</i></td>
    <td align="left" valign="middle">
<a href="https://github.com/stegua"><img src="https://github.com/stegua.png" width="40"></a>
</td>  </tr>
    <!-- <tr> <td align="left"><i>version</i></td> <td align="left"> </td> </tr> -->
    <tr> <td align="left"><i>Status</i></td> <td align="left">since 2020</td> </tr>
    <tr> <td align="left"><i>License</i></td> <td align="left"><a href="https://joinup.ec.europa.eu/sites/default/files/custom-page/attachment/2020-03/EUPL-1.2%20EN.txt">EUPL v1.2</a><i></i></td> </tr>
</table>

### <a name="References"></a>Main References

* **[1]** Bassetti F., Gualandi S., Veneroni M. [**On the computation of Kantorovich-Wasserstein distances between 2D-histograms by uncapacitated minimum cost flows**](https://epubs.siam.org/doi/abs/10.1137/19M1261195). SIAM J. Optim., 30(3), 2441–2469, 2020. Preprint on arXiv: [1804.00445](https://arxiv.org/abs/1804.00445).


### Note
Windows package of the R wrapper is built with [Win-Builder](https://win-builder.r-project.org/upload.aspx).


CHANGELOG.md
# Changelog

## [2022-12-09] Changelog from v0.4.1:

- Fix a compilation warning for CRAN (remove Italian letters with accent)


## [2021-05-07] Changelog from v0.4.0:

- New feature: support for focus areas.
- Fix a compilation error under Solaris 10. Ready for resubmission to CRAN
- Misc bug fixes.


## [2021-03-06] Changelog from v0.3.0:

- New feature: solve unbalanced optimal transport problems. See parameters "unbalanced" and "unbalanced_cost".
- New feature: solve optimal transport problem with or without precomputing the convex hull. See parameter "convex".
-Misc bug fixes.


## [2021-02-12] Changelog from v0.2.5:

- Fixing issues to the R wrapper raised by CRAN.
- Changing the "recode" option to TRUE by default (R wrapper)


## [2021-02-08] Changelog from v0.2.1 to v0.2.4:

- Fighting with the python wrapper and PyPI.


## [2021-01-28] Changelog v0.2.0:

- Fix several bugs to get ready for submission of the R wrapper to the CRAN repository


## [2021-01-13] Changelog v0.1.5:

- Improved Python wrapper with support to helper functions: compareOneToOne, compareOneToMany, compareAll
- New python example with numpy.array
- Docstring for python
- Bug fixing


## [2021-01-13] Changelog v0.1.4:

- Added three main helper functions
- Significantly improved the R wrapper, along with the documentation
- Added the column generation algorithm, with support for warm restarts
- Bug fixing


## [2020-01-01] Changelog v0.1.3:

- Internal private release for debugging


## [2020-12-05] Changelog v0.1.2:

- Added standalone CLI for the C++ library
- Fix a severe error in the computation of GCD (due to back-compatibility to C++11 instead of C++17)
- Added exact solver via complete bipartite graph
- Added log of progress via parameter
- Added support for internal runtime (to be compared with overall runtime)


## [2020-11-02] Changelog v0.1.1:

- new R example using "data.frame" objects
- fixed errors in c++ code
- added basic support for OpenMP (code is not yet full parallel)


wrappers/R/src/rcpp_module.cpp
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil;
// -*-
//
// rcpp_module.cpp: Rcpp R/C++ interface class library -- Rcpp Module examples
//
// Copyright (C) 2010 - 2012  Dirk Eddelbuettel and Romain Francois
//
// This file is part of Rcpp.
//
// Rcpp is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// Rcpp is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Rcpp.  If not, see <http://www.gnu.org/licenses/>.

#ifdef __GLIBC__
#define _POSIX_C_SOURCE 200809L
#endif

#include <Rcpp.h>

#include "KWD_Histogram2D.h"

RCPP_EXPOSED_AS(KWD::Histogram2D)

RCPP_MODULE(SKWD) {
	Rcpp::class_<KWD::Histogram2D>("Histogram2D")
		// expose the default constructor
		.constructor()
		//.constructor<int, int*, int*, double*>()

		.method("add", &KWD::Histogram2D::add, "add an non empty support point")
		.method("update", &KWD::Histogram2D::update,
			"update an non empty support point")
		.method("size", &KWD::Histogram2D::size,
			"return the number of nonempty points")
		.method("balance", &KWD::Histogram2D::balance,
			"return the total sum of all the weights")
		.method("normalize", &KWD::Histogram2D::normalize,
			"normalize the weights to sum them up to one");

	Rcpp::class_<KWD::Solver>("Solver")
		// expose the default constructor
		.constructor()

		// Solution methods
		.method("distance", &KWD::Solver::distance,
			"compute the distance between a pair of histograms with given L")

		.method("column_generation", &KWD::Solver::column_generation,
			"compute the distance between a pair of histograms with given L "
			"using column generation")

		.method("dense", &KWD::Solver::dense,
			"compute the distance between a pair of histograms with given L "
			"using a bipartite graph (slow on large instances)")

		// Paramaters and attributes
		.method("runtime", &KWD::Solver::runtime,
			"get the runtime in seconds of Network Simplex algorithm")

		.method("iterations", &KWD::Solver::iterations,
			"get the number of iterations of Network Simplex algorithm")

		.method("num_arcs", &KWD::Solver::num_arcs,
			"get the number of arcs in the Network model")

		.method("num_nodes", &KWD::Solver::num_nodes,
			"get the number of nodes in the Network model")

		.method("status", &KWD::Solver::status,
			"get the status of Network Simplex solver")

		.method("setDblParam", &KWD::Solver::setDblParam,
			"set a double parameter of the Network Simplex solver")

		.method("getDblParam", &KWD::Solver::getDblParam,
			"get a double parameter of the Network Simplex solver")

		.method("setStrParam", &KWD::Solver::setStrParam,
			"set a string parameter of the Network Simplex solver")

		.method("getStrParam", &KWD::Solver::getStrParam,
			"get a string parameter of the Network Simplex solver");
}

// [[Rcpp::export(compareOneToOne)]]
Rcpp::List compareOneToOne(
	Rcpp::NumericMatrix Coordinates, Rcpp::NumericMatrix Weights, int L = 3,
	bool recode = true, const std::string& method = "approx",
	const std::string& algorithm = "colgen",
	const std::string& model = "mincostflow",
	const std::string& verbosity = "silent", double timelimit = 14400,
	double opt_tolerance = 1e-06, bool unbalanced = false,
	double unbal_cost = 1e+09, bool convex = true) {
	if (Coordinates.ncol() != 2)
		throw(Rcpp::exception(
			"The Coordinates matrix must contain two columns for Xs and Ys."));

	if (Weights.ncol() < 2)
		throw(Rcpp::exception(
			"The Weigths matrix must contain two columns for W1 and W1."));

	if (Weights.ncol() > 2)
		Rprintf("WARNING: only the first two columns of matrix Weights are used as "
			"histograms.");

	// Input data
	int n = Coordinates.nrow();

	std::vector<int> data1 = Rcpp::as<std::vector<int>>(Coordinates);
	int* Xs = &data1[0];
	int* Ys = &data1[n];
	std::vector<double> data2 = Rcpp::as<std::vector<double>>(Weights);
	double* W1 = &data2[0];
	double* W2 = &data2[n];

	// Elaborate input parameters
	int LL = 3;
	if (L < 1)
		Rprintf("WARNING: Paramater L can take only value greater than 1. Using "
			"default value L=3.");
	else
		LL = L;

	KWD::Solver s;
	s.setStrParam(KWD_PAR_METHOD, method);
	s.setStrParam(KWD_PAR_MODEL, model);
	s.setStrParam(KWD_PAR_ALGORITHM, algorithm);
	s.setStrParam(KWD_PAR_VERBOSITY, verbosity);
	s.setDblParam(KWD_PAR_OPTTOLERANCE, opt_tolerance);
	s.setDblParam(KWD_PAR_TIMELIMIT, timelimit);
	if (recode)
		s.setStrParam(KWD_PAR_RECODE, KWD_VAL_TRUE);

	if (unbalanced) {
		s.setStrParam(KWD_PAR_UNBALANCED, KWD_VAL_TRUE);
		s.setDblParam(KWD_PAR_UNBALANCED_COST, unbal_cost);
	}

	if (convex)
		s.setStrParam(KWD_PAR_CONVEXHULL, KWD_VAL_TRUE);

	try {
		double d = -1;
		if (method == KWD_VAL_APPROX) {
			Rprintf("CompareOneToOne, Solution method: APPROX\n");
			d = s.compareApprox(n, Xs, Ys, W1, W2, LL);
		}
		else {
			Rprintf("CompareOneToOne, Solution method: EXACT\n");
			d = s.compareExact(n, Xs, Ys, W1, W2);
		}
		return Rcpp::List::create(
			Rcpp::Named("distance") = d, Rcpp::Named("runtime") = s.runtime(),
			Rcpp::Named("iterations") = s.iterations(),
			Rcpp::Named("nodes") = s.num_nodes(),
			Rcpp::Named("arcs") = s.num_arcs(), Rcpp::Named("status") = s.status());
	}
	catch (std::exception& e) {
		Rprintf("Error 13: Rcpp::NumericVector compareOneToOne()\n");
		forward_exception_to_r(e);
	}
	return Rcpp::List::create(Rcpp::Named("status") = "FAILED");
}

// [[Rcpp::export(compareOneToMany)]]
Rcpp::List compareOneToMany(
	Rcpp::NumericMatrix Coordinates, Rcpp::NumericMatrix Weights, int L = 3,
	bool recode = true, const std::string& method = "approx",
	const std::string& algorithm = "colgen",
	const std::string& model = "mincostflow",
	const std::string& verbosity = "silent", double timelimit = 14400,
	double opt_tolerance = 1e-06, bool unbalanced = false,
	double unbal_cost = 1e+09, bool convex = true) {
	if (Coordinates.ncol() != 2)
		throw(Rcpp::exception(
			"The Coordinates matrix must contain two columns for Xs and Ys."));

	if (Weights.ncol() < 2)
		throw(Rcpp::exception(
			"The Weights matrix must contain at least two columns."));

	// Input data
	int n = Coordinates.nrow();
	int m = Weights.ncol() - 1;

	std::vector<int> data1 = Rcpp::as<std::vector<int>>(Coordinates);
	int* Xs = &data1[0];
	int* Ys = &data1[n];

	std::vector<double> data2 = Rcpp::as<std::vector<double>>(Weights);
	double* W1 = &data2[0];
	double* Ws = &data2[n];

	// Elaborate input parameters
	int LL = 3;
	if (L < 1)
		Rprintf("WARNING: Paramater L can take only value greater than 1. Using "
			"default value L=3.");
	else
		LL = L;

	KWD::Solver s;
	s.setStrParam(KWD_PAR_METHOD, method);
	s.setStrParam(KWD_PAR_MODEL, model);
	s.setStrParam(KWD_PAR_ALGORITHM, algorithm);
	s.setStrParam(KWD_PAR_VERBOSITY, verbosity);
	s.setDblParam(KWD_PAR_OPTTOLERANCE, opt_tolerance);
	s.setDblParam(KWD_PAR_TIMELIMIT, timelimit);
	if (recode)
		s.setStrParam(KWD_PAR_RECODE, KWD_VAL_TRUE);

	if (unbalanced) {
		s.setStrParam(KWD_PAR_UNBALANCED, KWD_VAL_TRUE);
		s.setDblParam(KWD_PAR_UNBALANCED_COST, unbal_cost);
	}

	if (convex)
		s.setStrParam(KWD_PAR_CONVEXHULL, KWD_VAL_TRUE);

	try {
		Rcpp::NumericVector ds;
		if (method == KWD_VAL_APPROX) {
			Rprintf("CompareOneToMany, Solution method: APPROX\n");
			vector<double> _ds = s.compareApprox(n, m, Xs, Ys, W1, Ws, LL);
			for (auto v : _ds)
				ds.push_back(v);
		}
		else {
			Rprintf("CompareOneToMany, Solution method: EXACT\n");
			vector<double> _ds = s.compareApprox(n, m, Xs, Ys, W1, Ws, n - 1);
			for (auto v : _ds)
				ds.push_back(v);
		}
		return Rcpp::List::create(
			Rcpp::Named("distance") = ds, Rcpp::Named("runtime") = s.runtime(),
			Rcpp::Named("iterations") = s.iterations(),
			Rcpp::Named("nodes") = s.num_nodes(),
			Rcpp::Named("arcs") = s.num_arcs(), Rcpp::Named("status") = s.status());
	}
	catch (std::exception& e) {
		Rprintf("Error 13: Rcpp::NumericVector compareOneToMany()\n");
		forward_exception_to_r(e);
	}
	return Rcpp::List::create(Rcpp::Named("status") = "FAILED");
}

// [[Rcpp::export(compareAll)]]
Rcpp::List compareAll(Rcpp::NumericMatrix Coordinates,
	Rcpp::NumericMatrix Weights, int L = 3,
	bool recode = true, const std::string& method = "approx",
	const std::string& algorithm = "colgen",
	const std::string& model = "mincostflow",
	const std::string& verbosity = "silent",
	double timelimit = 14400, double opt_tolerance = 1e-06,
	bool unbalanced = false, double unbal_cost = 1e+09,
	bool convex = true) {
	if (Coordinates.ncol() != 2)
		throw(Rcpp::exception(
			"The Coordinates matrix must contain two columns for Xs and Ys."));

	if (Weights.ncol() < 2)
		throw(Rcpp::exception(
			"The Weights matrix must contain at least two columns."));

	// Input data
	int n = Coordinates.nrow();
	int m = Weights.ncol();

	std::vector<int> data1 = Rcpp::as<std::vector<int>>(Coordinates);
	int* Xs = &data1[0];
	int* Ys = &data1[n];

	std::vector<double> data2 = Rcpp::as<std::vector<double>>(Weights);
	double* Ws = &data2[0];

	// Elaborate input parameters
	int LL = 3;
	if (L < 1)
		Rprintf("WARNING: Paramater L can take only value greater than 1. Using "
			"default value L=3.");
	else
		LL = L;

	KWD::Solver s;
	s.setStrParam(KWD_PAR_METHOD, method);
	s.setStrParam(KWD_PAR_MODEL, model);
	s.setStrParam(KWD_PAR_ALGORITHM, algorithm);
	s.setStrParam(KWD_PAR_VERBOSITY, verbosity);
	s.setDblParam(KWD_PAR_OPTTOLERANCE, opt_tolerance);
	s.setDblParam(KWD_PAR_TIMELIMIT, timelimit);
	if (recode)
		s.setStrParam(KWD_PAR_RECODE, KWD_VAL_TRUE);

	if (unbalanced) {
		s.setStrParam(KWD_PAR_UNBALANCED, KWD_VAL_TRUE);
		s.setDblParam(KWD_PAR_UNBALANCED_COST, unbal_cost);
	}

	if (convex)
		s.setStrParam(KWD_PAR_CONVEXHULL, KWD_VAL_TRUE);

	try {
		Rcpp::NumericMatrix ds;
		if (method == KWD_VAL_APPROX) {
			Rprintf("CompareAll, Solution method: APPROX\n");
			vector<double> _ds = s.compareApprox(n, m, Xs, Ys, Ws, LL);
			for (auto v : _ds)
				ds.push_back(v);
		}
		else {
			Rprintf("CompareAll, Solution method: EXACT\n");
			vector<double> _ds = s.compareApprox(n, m, Xs, Ys, Ws, n - 1);
			for (auto v : _ds)
				ds.push_back(v);
		}
		return Rcpp::List::create(
			Rcpp::Named("distance") = ds, Rcpp::Named("runtime") = s.runtime(),
			Rcpp::Named("iterations") = s.iterations(),
			Rcpp::Named("nodes") = s.num_nodes(),
			Rcpp::Named("arcs") = s.num_arcs(), Rcpp::Named("status") = s.status());
	}
	catch (std::exception& e) {
		Rprintf("Error 13: Rcpp::NumericVector compareAll()\n");
		forward_exception_to_r(e);
	}
	return Rcpp::List::create(Rcpp::Named("status") = "FAILED");
}

// [[Rcpp::export(focusArea)]]
Rcpp::List focusArea(Rcpp::NumericMatrix Coordinates,
	Rcpp::NumericMatrix Weights, int x, int y, int radius,
	int L = 3, bool recode = true,
	const std::string& method = "approx",
	const std::string& algorithm = "colgen",
	const std::string& model = "mincostflow",
	const std::string& verbosity = "silent",
	double timelimit = 14400, double opt_tolerance = 1e-06,
	const std::string& area = "l2") {
	if (Coordinates.ncol() != 2)
		throw(Rcpp::exception(
			"The Coordinates matrix must contain two columns for Xs and Ys."));

	if (Weights.ncol() < 2)
		throw(Rcpp::exception(
			"The Weigths matrix must contain two columns for W1 and W1."));

	if (Weights.ncol() > 2)
		Rprintf("WARNING: only the first two columns of matrix Weights are used as "
			"histograms.");

	// Input data
	int n = Coordinates.nrow();

	std::vector<int> data1 = Rcpp::as<std::vector<int>>(Coordinates);
	int* Xs = &data1[0];
	int* Ys = &data1[n];
	std::vector<double> data2 = Rcpp::as<std::vector<double>>(Weights);
	double* W1 = &data2[0];
	double* W2 = &data2[n];

	// Elaborate input parameters
	int LL = 3;
	if (L < 1)
		Rprintf("WARNING: Paramater L can take only value greater than 1. Using "
			"default value L=3.");
	else
		LL = L;

	KWD::Solver s;
	s.setStrParam(KWD_PAR_METHOD, method);
	s.setStrParam(KWD_PAR_MODEL, model);
	s.setStrParam(KWD_PAR_ALGORITHM, algorithm);
	s.setStrParam(KWD_PAR_VERBOSITY, verbosity);
	s.setDblParam(KWD_PAR_OPTTOLERANCE, opt_tolerance);
	s.setDblParam(KWD_PAR_TIMELIMIT, timelimit);
	if (recode)
		s.setStrParam(KWD_PAR_RECODE, KWD_VAL_TRUE);
	s.setStrParam(KWD_PAR_AREA, area);

	try {
		Rprintf("FocusArea, Solution method: APPROX\n");

		double d = s.focusArea(n, Xs, Ys, W1, W2, x, y, radius, LL);

		return Rcpp::List::create(
			Rcpp::Named("distance") = d, Rcpp::Named("runtime") = s.runtime(),
			Rcpp::Named("iterations") = s.iterations(),
			Rcpp::Named("nodes") = s.num_nodes(),
			Rcpp::Named("arcs") = s.num_arcs(), Rcpp::Named("status") = s.status());
	}
	catch (std::exception& e) {
		Rprintf("Error 13: Rcpp::NumericVector FocusArea()\n");
		forward_exception_to_r(e);
	}

	return Rcpp::List::create(Rcpp::Named("status") = "FAILED");
}
